exp/addons/controllers/clusterresourceset_controller_test.go

```diff
@@ -236,7 +236,80 @@ metadata:
 			return apierrors.IsNotFound(err)
 		}, timeout).Should(BeTrue())
 	})
-	It("Should reconcile a ClusterResourceSet when a resource is created that is part of ClusterResourceSet resources", func() {
+	It("Should reconcile a ClusterResourceSet when a Secret resource is created that is part of ClusterResourceSet resources", func() {
+		labels := map[string]string{"foo2": "bar2"}
+		newSecretName := "test-secret3"
+
+		clusterResourceSetInstance := &addonsv1.ClusterResourceSet{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "test-clusterresourceset",
+				Namespace: defaultNamespaceName,
+			},
+			Spec: addonsv1.ClusterResourceSetSpec{
+				ClusterSelector: metav1.LabelSelector{
+					MatchLabels: labels,
+				},
+				Resources: []addonsv1.ResourceRef{{Name: newSecretName, Kind: "Secret"}},
+			},
+		}
+		// Create the ClusterResourceSet.
+		Expect(testEnv.Create(ctx, clusterResourceSetInstance)).To(Succeed())
+
+		testCluster.SetLabels(labels)
+		Expect(testEnv.Update(ctx, testCluster)).To(Succeed())
+
+		By("Verifying ClusterResourceSetBinding is created with cluster owner reference")
+		// Wait until ClusterResourceSetBinding is created for the Cluster
+		clusterResourceSetBindingKey := client.ObjectKey{
+			Namespace: testCluster.Namespace,
+			Name:      testCluster.Name,
+		}
+		Eventually(func() bool {
+			binding := &addonsv1.ClusterResourceSetBinding{}
+
+			err := testEnv.Get(ctx, clusterResourceSetBindingKey, binding)
+			return err == nil
+		}, timeout).Should(BeTrue())
+
+		// Initially ConfigMap is missing, so no resources in the binding.
+		Eventually(func() bool {
+			binding := &addonsv1.ClusterResourceSetBinding{}
+
+			err := testEnv.Get(ctx, clusterResourceSetBindingKey, binding)
+			if err == nil {
+				if len(binding.Spec.Bindings) > 0 && len(binding.Spec.Bindings[0].Resources) == 0 {
+					return true
+				}
+			}
+			return false
+		}, timeout).Should(BeTrue())
+
+		time.Sleep(2 * time.Second)
```

Comment:
What is the a reason why we are waiting here? Or the eventually statement statement at line 299 is enough to wait for the controllers to catch up?

Reply:
Needed to make sure the previous reconcile is over before creating a Secret resource, because if I don't add sleep here what sometimes happens is that the resource is created somewhere in the middle of the reconcile and the resource is reconciled within that reconcile. Hence we do not test if predicates work. Sleep is to make sure the previous reconciles are over and when a resource is created the watch on Secrets will kick.

Already have a comment in the main branch, now added here too.